### PR TITLE
CI: add GCC 11, GCC 12, Clang 13, Clang 14

### PR DESCRIPTION
**Pull request by thesamesam**
Notably, this might have caught #18225, as Clang 14 wasn't - and is not yet until this PR - in OpenSSL's CI.

It makes sense to ensure CI tests compilers used in newer Linux distributions:
- Fedora 36 ships with GCC 12
- Ubuntu 22.04 ships with Clang 14

### CODE REVIEW

File: .github/workflows/compiler-zoo.yml [hunk 1]
```diff
@@ -16,33 +16,72 @@ jobs:
...
+        echo "deb-src http://apt.llvm.org/jammy/ llvm-toolchain-jammy-13 main" | sudo tee /etc/apt/sources.list.d/llvm.list
+        echo "deb http://apt.llvm.org/jammy/ llvm-toolchain-jammy-14 main" | sudo tee /etc/apt/sources.list.d/llvm.list
+        echo "deb-src http://apt.llvm.org/jammy/ llvm-toolchain-jammy-14 main" | sudo tee /etc/apt/sources.list.d/llvm.list
+
+        echo "deb https://ppa.launchpadcontent.net/ubuntu-toolchain-r/ppa/ubuntu jammy main" | sudo tee /etc/apt/sources.list.d/gcc.list
+        echo "deb-src https://ppa.launchpadcontent.net/ubuntu-toolchain-r/ppa/ubuntu jammy main" | sudo tee /etc/apt/sources.list.d/gcc.list
+
+        # From https://launchpad.net/~ubuntu-toolchain-r/+archive/ubuntu/ppa
+        sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 60C317803A41BA51845E371A1E9377A2BA9EF27F
+
```
Comment (t8m) [inline .github/workflows/compiler-zoo.yml:84]: Why is this needed? Or, for which clang/gcc versions this is needed?

Comment (thesamesam) [reply .github/workflows/compiler-zoo.yml]: Checked again and it's a left over experiment from trying to make it all work with Ubuntu 20.04. Dropped as nothing needs it now.

In theory, we could keep the Clang one to get latest 13.x and 14.x, but let's not worry about it for now.

**Review by paulidale — APPROVED** [review]
Nice work

**Review by t8m — APPROVED** [review]

**Comment by thesamesam** [comment]
Before we merge, we might need to look at https://github.com/thesamesam/openssl/runs/7032755752?check_suite_focus=true#step:6:5 for the Clang 14 failure. The GCC 12 failure was already addressed by @t8m in #18627 (thanks!)

The Clang failure is:
```
test/test_test.c:58:9: note: cast one or both operands to int to silence this warning
test/test_test.c:58:9: error: use of bitwise '|' with boolean operands [-Werror,-Wbitwise-instead-of-logical]
    if (!TEST(1, TEST_uint_eq(3u, 3u))
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
test/test_test.c:58:9: note: cast one or both operands to int to silence this warning
```

How do we want to handle it?

**Comment by paulidale** [comment]
Fixing test_test.c is easy -- change all the `|` to `||`.

I used `|` originally so we'd see multiple errors if there were problems but it's not critical.

**Comment by thesamesam** [comment]
Okay, happy to throw that in here (or you may of course do it in a separate PR) if it's fine with you. I was reluctant to do it myself because I didn't get why it was single-`|` in the first place, but now I know!

**Review by paulidale — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by paulidale** [comment]
Merged, thanks for this test improvement.

**Comment by paulidale** [comment]
I just noticed that these commits have `CLA: trivial`, they are well beyond what we'd consider trivial.
@thesamesam, could you please submit a [completed ICLA](https://www.openssl.org/policies/openssl_icla.pdf)?

**Comment by thesamesam** [comment]
> I just noticed that these commits have `CLA: trivial`, they are well beyond what we'd consider trivial. @thesamesam, could you please submit a [completed ICLA](https://www.openssl.org/policies/openssl_icla.pdf)?

Happily! Let me do that now.

**Pull request closed without merging**